# barapost toolkit
**So to speak, "*a posteriori* barcoding"**

## 1. barapost.py

### DESCRIPTION:
 Script "barapost.py" is designed for determinating the taxonomic position
    of long reads by blasting each of them and regarding the best hit.

 Script processes FASTQ (as well as '.fastq.gz') files.

 Results of the work of this script are written to TSV file,
    that can be found in result directory.

 If no separate FASTQ files and not input directory is specified,
    "barapost.by" will process all FASTQ files in current directory.

 FASTQ files processed by this script are meant to be sorted afterwards by "fastq_sorted.py".


### OPTIONS:
```
    -h (--help) --- show help message;

    -f (--infile) --- input FASTQ (or '.fastq.gz') file.
            You can specify multiple input files with this option (see EXAMPLES #2);

    -d (--indir) --- directory which contains FASTQ (or '.fastq.gz') files meant to be processed.
            E.i. all FASTQ files in this direcory will be processed;

    -o (--outdir) --- output directory;

    -p (--packet-size) --- size of the packet, e.i. number of sequence to blast in one request.
            Value: integer number [1, 1000]. Default value is 20;

    -a (--algorithm) --- BLASTn algorithm to use for aligning.
            Available values: 'megaBlast', 'discoMegablast', 'blastn'.
            Default is megaBlast;

    -g (--organisms) --- 'nt' database restriction, e.i. organisms that you expect to see in result files.
            Format of value: 
              <organism1_name>,<organism1_taxid>+<organism2_name>,<organism2_taxid>+...
            See EXAMPLES #2 and #3 below.
            Spaces are not allowed. Number of organisms can be from 1 to 5 (5 is maximum).
            Default value is:   bacteria,2+viruses,10239.
```

- More clearly, functionality of '-g' this option is totally equal to "Organism" text boxes on this BLASTn page:
    'https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome'.
- You can find your Taxonomy IDs here: 'https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi'.

### EXAMPLES:
  1) Process one file with default settings:

`./barapost.py -f reads.fastq`

  2) Process two files with discoMegablast, packet size of 5 reads.
       Search only among Erwinia sequences:

`./barapost.py -f reads_1.fastq.gz -f reads_2.fastq -a discoMegablast -p 5 -g Erwinia,551`

  3) Process all FASTQ (and '.fastq.gz') files in directory named "some_dir".
     Search only among Escherichia and viral sequences:

`./barapost.py -d some_dir -g Escherichia,561+viruses,10239 -o outdir`


## 2. fastq_sorter.py

### DESCRIPTION:
 Script "fastq_forter.py" is designed for sorting FASTQ files processed by "barapost.py".

 Requires result TSV file generated by "barapost.py" and FASTQ files processed by it.
 Therefore '-b' option is mandatory.

 Separate FASTQ files should be specified with -f(--infile) option (see EXAMPLES #2 below).
 
 If -d (--indir) option is specified, all FASTQ files in directory specified by this option will be sorted.
 
 If no separate FASTQ files and not input directory is specified,
       "fastq_sorter.py" will process all FASTQ files in current directory.


### OPTIONS (* means mandatory option):
```
    -h (--help) --- show help message;

  * -b (--barapost-result) --- result TSV file generated by "barapost.py";

    -f (--infile) --- input FASTQ file (can be gzipped);

    -d (--indir) --- directory which contains FASTQ files (files can be gzipped) meant to be sorted;

    -o (--outdir) --- output directory;

    -s (--sorting-sensitivity) --- sorting sensitivity;
        E.i. the lowest taxonomy rank that will be used in names of resut files;
        Available values: 'genus', 'species', 'strain';
```

### EXAMPLES:

  1) Process one FASTQ file with default settings:

`./fastq_sorter.py -f reads.fastq.gz -b test_megaBlast_result.tsv`

  2) Process two files, place results in "outdir" directory:

`./fastq_sorter.py -f reads_1.fastq.gz -f reads_2.fastq -o outdir -b test_megaBlast_result.tsv`

  3) Process all files in directory named "dir_with_reads". Sort by genus:

`./fastq_sorter.py -d dir_with_reads -o outdir -b test_megaBlast_result.tsv -s genus`

## Example of usage in combination:

You can try this scripts on test dataset named "some_reads.fastq" (there are 4 reads):

  `./barapost.py -f some_reads.fastq -o ./ -g Escherichia,561+viruses,10239`

  `./fastq_sorter.py -f some_reads.fastq -b some_reads/some_reads_megaBlast_result.tsv -o some_sorted_reads`
